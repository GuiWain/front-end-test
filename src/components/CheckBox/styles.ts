import styled, { css, DefaultTheme } from 'styled-components'

import { CheckBoxProps } from '.'

export type WrapperProps = {
  error?: boolean
} & Pick<CheckBoxProps, 'disabled'>

const wrapperModifiers = {
  error: (theme: DefaultTheme) => css`
    ${Label} {
      color: ${theme.colors.errorMedium};
    }
  `,
  disabled: (theme: DefaultTheme) => css`
    ${Label}, ${CheckBox} {
      cursor: not-allowed;
      filter: saturate(30%);
      color: ${theme.colors.neutralDark};
    }
  `
}

export const Wrapper = styled.div<WrapperProps>`
  ${({ theme, error, disabled }) => css`
    ${error && wrapperModifiers.error(theme)}
    ${disabled && wrapperModifiers.disabled(theme)}
  `}
`

export const InputCheckBox = styled.div`
  display: flex;
  gap: 0.5rem;
  align-items: center;

  width: auto;
`

export const CheckBox = styled.input.attrs({
  type: 'checkbox',
  'data-testid': 'CheckBox'
})`
  justify-self: flex-end;
  width: 1.5rem;
  height: 1.5rem;
  cursor: pointer;
`

export const Label = styled.label`
  ${({ theme }) => css`
    color: ${theme.colors.secondaryDark};
    font: normal normal ${theme.font.regular} ${theme.font.sizes.xxsmall} /
      2.4rem ${theme.font.family};
    letter-spacing: 0rem;

    cursor: pointer;
    margin: 0;

    justify-self: flex-start;
  `}
`

export const Error = styled.p`
  ${({ theme }) => css`
    color: ${theme.colors.errorLight};
    font-size: ${theme.font.sizes.xxxsmall};
    margin: 0.5rem 0;
  `}
`
