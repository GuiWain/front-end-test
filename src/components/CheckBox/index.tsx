import { useState } from 'react'
import * as S from './styles'

export type CheckBoxProps = {
  getValues?: (field: boolean) => void
  label?: string
  name?: string
  id?: string
  value?: string
  disabled?: boolean
  defaultChecked?: boolean
  required?: boolean
  error?: string
}

const CheckBox = ({
  getValues,
  label,
  value,
  disabled = false,
  defaultChecked = false,
  error,
  name,
  id,
  required = false
}: CheckBoxProps) => {
  const [checked, setChecked] = useState(
    defaultChecked ? defaultChecked : false
  )

  const onChange = (value: boolean) => {
    setChecked(value)
    !!getValues && getValues(value)
  }

  return (
    <S.Wrapper error={!!error} disabled={disabled}>
      <S.InputCheckBox>
        <S.CheckBox
          value={value}
          disabled={disabled}
          checked={checked}
          name={name}
          id={name ? name : id}
          onChange={(e) => onChange(e.target.checked)}
          required={required}
        />
        {!!label && <S.Label htmlFor={name}>{label}</S.Label>}
      </S.InputCheckBox>
      {!!error && <S.Error>{error}</S.Error>}
    </S.Wrapper>
  )
}

export default CheckBox
