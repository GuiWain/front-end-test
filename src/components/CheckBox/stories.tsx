import { Story, Meta } from '@storybook/react'
import CheckBox, { CheckBoxProps } from '.'

export default {
  title: 'inputs/CheckBox',
  component: CheckBox,
  argTypes: {
    label: {
      type: 'string'
    },
    value: {
      type: 'string'
    },
    defaultChecked: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    },
    disabled: {
      control: {
        type: 'radio',
        options: [true, false]
      }
    },
    name: {
      type: 'string'
    }
  }
} as Meta

export const Default: Story<CheckBoxProps> = (args) => <CheckBox {...args} />

Default.args = {
  label: 'Checkbox Label.',
  value: 'chkbx',
  name: 'chkbx',
  defaultChecked: false,
  disabled: false
}

export const WithError: Story<CheckBoxProps> = (args) => <CheckBox {...args} />

WithError.args = {
  label: 'Checkbox Label..',
  value: 'chkbx',
  name: 'chkbx',
  defaultChecked: false,
  disabled: false,
  error: 'Field required'
}
