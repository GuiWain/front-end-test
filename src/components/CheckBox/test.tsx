import userEvent from '@testing-library/user-event'
import { render, screen } from 'utils/test-utils'

import CheckBox from '.'

describe('<CheckBox />', () => {
  it('should render the CheckBox', () => {
    const { container } = render(<CheckBox />)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render the CheckBox with Label', () => {
    render(<CheckBox label="Checkbox" name="name" />)

    expect(screen.getByText('Checkbox')).toBeInTheDocument()
  })

  it('should render the CheckBox checked', () => {
    render(<CheckBox />)

    const input = screen.getByTestId('CheckBox')
    userEvent.click(input)

    expect(input).toBeChecked()
  })

  it('should render the CheckBox with getValues', () => {
    const getValues = jest.fn()
    render(<CheckBox getValues={getValues} />)

    const input = screen.getByTestId('CheckBox')
    userEvent.click(input)

    expect(getValues).toHaveBeenCalled()
  })

  it('should render the CheckBox with error', () => {
    render(<CheckBox error="Required field" />)

    const input = screen.getByTestId('CheckBox')
    userEvent.click(input)

    expect(screen.getByText('Required field')).toBeInTheDocument()
  })

  it('should render the CheckBox disabled', () => {
    render(<CheckBox disabled />)

    const input = screen.getByTestId('CheckBox')
    expect(input).toBeDisabled()
  })

  it('should render the CheckBox width default defaultChecked', () => {
    render(<CheckBox defaultChecked={true} />)

    const input = screen.getByTestId('CheckBox')
    expect(input).toBeChecked()
  })
})
