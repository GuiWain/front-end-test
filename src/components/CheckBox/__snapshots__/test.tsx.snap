// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckBox /> should render the CheckBox 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
}

.c1 {
  justify-self: flex-end;
  width: 1.5rem;
  height: 1.5rem;
  cursor: pointer;
}

<div
  class=""
>
  <div
    class="c0"
  >
    <input
      class="c1"
      data-testid="CheckBox"
      type="checkbox"
      value=""
    />
  </div>
</div>
`;
